core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java

```diff
@@ -0,0 +1,249 @@
+    this.outputTempFile = Utils.tempFileWith(outputFile);
+  }
+
+  @Override
+  public ShufflePartitionWriter getNextPartitionWriter() throws IOException {
+    initStream();
+    initChannel();
+    return new DefaultShufflePartitionWriter(currPartitionId++);
+  }
+
+  @Override
+  public void commitAllPartitions() throws IOException {
+    cleanUp();
+    blockResolver.writeIndexFileAndCommit(shuffleId, mapId, partitionLengths, outputTempFile);
+    if (!successfulWrite) {
+      if (!outputFile.getParentFile().isDirectory() && !outputFile.getParentFile().mkdirs()) {
+        throw new IOException(
+          String.format(
+            "Failed to create shuffle file directory at %s.",
+            outputFile.getParentFile().getAbsolutePath()));
+      }
+      if (!outputFile.isFile() && !outputFile.createNewFile()) {
+        throw new IOException(
+          String.format(
+            "Failed to create empty shuffle file at %s.", outputFile.getAbsolutePath()));
+      }
+    }
+  }
+
+  @Override
+  public void abort(Throwable error) throws IOException {
+    if (!outputTempFile.delete() && outputTempFile.exists()) {
+      log.warn("Failed to delete temporary shuffle file at {}", outputTempFile.getAbsolutePath());
+    }
+    if (!outputFile.delete() && outputFile.exists()) {
+      log.warn("Failed to delete outputshuffle file at {}", outputFile.getAbsolutePath());
+    }
+    cleanUp();
+  }
+
+  private void cleanUp() throws IOException {
+    if (outputBufferedFileStream != null) {
+      outputBufferedFileStream.close();
+    }
+
+    if (outputFileStream != null) {
+      outputFileStream.close();
+    }
+  }
+
+  private void initStream() throws IOException {
+    if (outputFileStream == null) {
+      outputFileStream = new FileOutputStream(outputTempFile, true);
+      ts = new TimeTrackingOutputStream(metrics, outputFileStream);
+    }
+    if (outputBufferedFileStream == null) {
+      outputBufferedFileStream = new BufferedOutputStream(ts, bufferSize);
+    }
+  }
+
+  private void initChannel() throws IOException {
+    if (outputFileStream == null) {
+      outputFileStream = new FileOutputStream(outputTempFile, true);
+    }
+    if (outputFileChannel == null) {
+      outputFileChannel = outputFileStream.getChannel();
+    }
+  }
+
+  private class DefaultShufflePartitionWriter implements ShufflePartitionWriter {
+
+    private final int partitionId;
+    private DefaultShuffleBlockOutputStream stream = null;
+    private DefaultShuffleBlockByteChannel channel = null;
+
+    private DefaultShufflePartitionWriter(int partitionId) {
+      this.partitionId = partitionId;
+    }
+
+    @Override
+    public OutputStream openStream() throws IOException {
+      stream = new DefaultShuffleBlockOutputStream();
+      return stream;
+    }
+
+    @Override
+    public long getLength() {
+      if (channel != null && stream == null) {
+        try {
+          channel.close();
```

Comment:
This will close the underlying file before we're done writing to it for subsequent partitions. This should be verifiable by running a unit test - `BypassMergeSortShuffleWriterSuite` - using multiple partitions.

Reply:
Actually never mind sorry - just noticed that `channel` isn't the file channel but it's the shielded from close output channel.